### ISSUE TITLE
Translation Sub-Menu "sort notes by >" 

### ISSUE THREAD

**Issue by foxmask** [laurent22/joplin#572]
### Operating system

- Windows
- macOS
- Linux
- Android
- iOS

### Application

- Desktop
- Mobile
- Terminal

### Description

concerns 

* "Updated date"
* "Created date"
* "Title"

#### 1rst try

I tried to add the (french) translation for the sub menu "sort by notes" which displays :

By adding them to CliCli/locale/fr_FR.po

```
msgid "Updated date"
msgstr "Date mise à jour"
```

But the build process erases the `joplin.pot` file and recreate it from the strings (surrounded by `_()`) it found in the source.

But as the strings come from calulated object, the translation process never found the string sand then never get the strings at all.

I spotted those string here :

```
ElectronClient/app/lib/models/Note.js:			user_updated_time: 'updated date',
ReactNativeClient/lib/models/Note.js:			user_updated_time: 'updated date',
```

the function is

```
        static fieldToLabel(field) {
                const fieldsToLabels = {
                        title: 'title',
                        user_updated_time: 'updated date',
                        user_created_time: 'created date',
                };

                return field in fieldsToLabels ? fieldsToLabels[field] : field;
        }
```

then `fieldToLabel` is used in a `toTitleCase()` function:
```
options[noteSortFields[i]] = toTitleCase(Note.fieldToLabel(noteSortFields[i]));
```
even with `_()` in 

```
options[noteSortFields[i]] = toTitleCase(_(Note.fieldToLabel(noteSortFields[i])));
```

does not do the trick.

#### 2nd try in `fieldToLabel`

I also try to use _()

```
        static fieldToLabel(field) {
                const fieldsToLabels = {
                        title: _('title'),
                        user_updated_time: _('updated date'),
                        user_created_time: _('created date'),
                };

                return field in fieldsToLabels ? fieldsToLabels[field] : field;
        }
```
then when building the translation, the fuzzy function translate that correctly and then comments them with '# ~ " 

```
# fuzzy
#~ msgid "updated date"
#~ msgstr "date mise à jour"
```

#### conclusion
I wanted to contribute but I think I reached my limit :) I feel that the build process of translation, in that specific case, does not allow to translate those string, but i'm not sure

**Comment by laurent22** [comment]
@foxmask, for a string to be translatable, it needs to be wrapped in `_()`, like you did in your second try. Then you can run build-translation.js, and that will add the string to the .pot file and all the .po files. After that it can be translated in each .po file.

I think the issue here is that you've added the string to the .po file before marking it with `_()`. I'd expect that if you now open the file in Poedit, you should see that string and be able to translate it.

**Comment by foxmask** [comment]
So, 
I tried again and change ReactNativeClient/lib/models/Note.js with the `_()` function.
then I did 
```
node Tools/build-translations.js
```
at that moment, the CliClient/locale/fr_FR.po is translated with 

```poedit
#, fuzzy
msgid "title"
msgstr "Sans titre"

#, fuzzy
msgid "updated date"
msgstr "Mis à jour : %d."

#, fuzzy
msgid "created date"
msgstr "Créés : %d."
```

then I did 

```
cd ElectronClient/app
rsync --delete -a ../../ReactNativeClient/lib/ lib/
npm install
yarn dist
```

then I checked the generated files 

```
find . -name fr_FR.json
./ReactNativeClient/locales/fr_FR.json
./ElectronClient/app/locales/fr_FR.json
./ElectronClient/app/dist/linux-unpacked/resources/app/locales/fr_FR.json
./CliClient/build/locales/fr_FR.json
```

all files are the same but contains: 

```
[...]
 nom réservé.","title":"title","updated date":"updated date","created date":"created date","Untitled":"Sans titre",
[...]
```
and when I start Joplin the sub menu is still not translated.

Did I miss somthing ?

**Comment by laurent22** [comment]
You would need to run build-translation.js again so that it creates the .json files

**Comment by foxmask** [comment]
That does not change anything, the json files did not change (same size) and still don't contains the translation I made in the fr_FR.po.
I will track what's the matter with the msgmerge to try to understand why the translated string are not merged inside the fr_FR.json files.

**Comment by foxmask** [comment]
Ok I did it !

process I wrote to remember the next time:

for the example, let's say we translate foobar to FooBar

1. add the `_()` where the translation is needed on `'foobar`' to become `_('foobar')`
2. execute `node Tools/build-translation.js` to make joplin spotting the translation and write it in the `CliClient/locale/fr_FR.po` file
3. edit `CliClient/locale/fr_FR.po` and look for `msgid 'foobar'`, and if '# fuzzy" comes before it, drop them too and add below `msgid 'foobar'` add msgstr "FooBar"
4. execute once again `node Tools/build-translation.js` for this time add the FooBar translation in each `*/locales/fr_FR.js` file (for example `ReactNativeClient/locales/fr_FR.json`,  `ElectronClient/app/locales/fr_FR.json`, `CliClient/locales/fr_FR.po`
5. generate the Joplin binary file:
```
cd ElectronClient/app
rsync --delete -a ../../ReactNativeClient/lib/ lib/
npm install
yarn dist
```
6. checkout your translations by running `./dist/Joplin-1.0.XX-x86_64.AppImage`
